SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestMultiServerUnrealComponents/SpatialTestMultiServerUnrealComponents.cpp

```diff
@@ -0,0 +1,169 @@
+
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDKSettings.h"
+#include "TestUnrealComponents.h"
+#include "TestUnrealComponentsActor.h"
+
+#include "Kismet/GameplayStatics.h"
+#include "Net/UnrealNetwork.h"
+
+/**
+ * This test tests that static and dynamic component are properly resolved on expectant workers.
+ *
+ * The test includes 2 Server and 2 Clients.
+ * The flow is as follows:
+ * - Setup:
+ *  - The TestActor is spawned with both static and dynamic components.
+ *  - The components are initialized with unique starting values.
+ *  - The TestActor is possessed by Client1
+ * - Test:
+ *	- The owner client checks it can see the actor, components, and the values are correct.
+ *  - The non-auth server checks it can see the actor, components, and the values are correct.
+ *  - The non-owning client checks it can see the actor, components, and the values are correct.
+ * - Clean-up:
+ *	- The TestActor is destroyed.
+ */
+ASpatialTestMultiServerUnrealComponents::ASpatialTestMultiServerUnrealComponents()
+	: Super()
+{
+	Author = "Mike";
+	Description = TEXT("Test Unreal Component Replication in a MultiServer Context");
+}
+
+void ASpatialTestMultiServerUnrealComponents::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ThisClass, TestActor);
+}
+
+void ASpatialTestMultiServerUnrealComponents::PrepareTest()
+{
+	Super::PrepareTest();
+
+	bInitialOnlyEnabled = GetDefault<USpatialGDKSettings>()->bEnableInitialOnlyReplicationCondition;
+
+	// The Server spawns the TestActor and immediately after it creates and attaches the dynamic components
+	AddStep(TEXT("SpatialTestMultiServerUnrealComponents Spawn and Setup Test Actor"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		if (bInitialOnlyEnabled)
+		{
+			AddExpectedLogError(TEXT("Dynamic component using InitialOnly data. This data will not be sent."), 1, false);
+		}
+
+		AssertTrue(HasAuthority(), TEXT("Server requires auth"));
+		TestActor = GetWorld()->SpawnActor<ATestUnrealComponentsActor>(ActorSpawnPosition, FRotator::ZeroRotator, FActorSpawnParameters());
+		TestActor->SpawnDynamicComponents();
+
+		const bool bWrite = true;
+		ProcessActorProperties(bWrite);
+
+		AController* PlayerController = Cast<AController>(GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1)->GetOwner());
+		AssertTrue(PlayerController->HasAuthority(), TEXT("Server requires authority over controller"));
+		TestActor->SetOwner(PlayerController);
+
+		RegisterAutoDestroyActor(TestActor);
+
+		FinishStep();
+	});
+
+	// Check on owning client that actor exists and properties are correct
+	AddStep(
+		TEXT("SpatialTestMultiServerUnrealComponents Validate Test Actor On Owning Client"), FWorkerDefinition::Client(1),
+		[this]() -> bool {
+			return TestActor != nullptr;
+		},
+		[this]() {
+			AssertFalse(TestActor->HasAuthority(), TEXT("This client shouldn't have authority"));
```

Comment:
I guess we can leave this in, even though the clients should never have authority.
What about a `AssertTrue(TestActor->IsOwnedBy(MyPlayer))` in addition?
But not really required, just thought it would be more appropriate than the authority check.

Reply:
They can if they're spawned locally! But yeah can add the additional IsOwned check.